File: 2021/12/29/main.go
package main

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	helper(root)
	return root
}

func helper(node *Node) {
	if node == nil {
		return
	}
	if node.Left != nil {
		node.Left.Next = node.Right
	}
	if node.Next != nil && node.Left != nil {
		node.Right.Next = node.Next.Left
	}
	helper(node.Left)
	helper(node.Right)
}

File: 2021/12/29/main.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        def helper(node):
            if not node: return None
            if node.left:
                node.left.next = node.right
            if node.next and node.left:
                node.right.next = node.next.left
            helper(node.left)
            helper(node.right)
        
        helper(root)
        return root